src/react/utils/useHydrateAtoms.ts

```diff
@@ -4,11 +4,32 @@ import type { WritableAtom } from '../../vanilla.ts'
 type Store = ReturnType<typeof useStore>
 type Options = Parameters<typeof useStore>[0]
 type AnyWritableAtom = WritableAtom<unknown, any[], any>
+type AtomMap<A = AnyWritableAtom, V = unknown> = Map<A, V>
+type AtomTuple<A = AnyWritableAtom, V = unknown> = readonly [A, V]
+type InferAtoms<T extends Iterable<AtomTuple>> = {
+  [K in keyof T]: T[K] extends AtomTuple<infer A>
+    ? A extends AnyWritableAtom
+      ? AtomTuple<A, ReturnType<A['read']>>
+      : T[K]
+    : never
+}
 
 const hydratedMap: WeakMap<Store, WeakSet<AnyWritableAtom>> = new WeakMap()
 
-export function useHydrateAtoms(
-  values: Iterable<readonly [AnyWritableAtom, unknown]>,
+export function useHydrateAtoms<T extends Array<AtomTuple>>(
+  values: InferAtoms<T>,
+  options?: Options
+): void
+export function useHydrateAtoms<T extends AtomMap>(
+  values: T,
+  options?: Options
+): void
+export function useHydrateAtoms<T extends Iterable<AtomTuple>>(
+  values: InferAtoms<T>,
+  options?: Options
+): void
```

Comment:
Added this because when I was trying to get Typescript to pick the correct overload by moving `export function useHydrateAtoms<T extends AtomMap>` as the function implementation signature I noticed the following error in the tests:
``
useHydrateAtoms.ts(27, 17): The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.
``
![Screenshot 2023-05-01 at 10 38 25 PM](https://user-images.githubusercontent.com/525506/235574399-42075309-7d93-480c-bfeb-2204075d55fb.png)